pkg/tls/reloader.go

```diff
@@ -0,0 +1,172 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tls
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+)
+
+// CertReloader is the struct that parses a certificate/key pair,
+// providing a goroutine safe GetCertificate method to retrieve the parsed content.
+//
+// The GetCertificate signature is compatible with https://golang.org/pkg/crypto/tls/#Config.GetCertificate
+// and can be used to hot-reload a certificate/key pair.
+//
+// For hot-reloading the Watch method must be started explicitly.
+type CertReloader struct {
+	certPath, keyPath string
+
+	// contains the watch targets (keys)
+	// and the watch locations (values)
+	watchables map[string]string
+
+	mu   sync.RWMutex // protects the fields below
+	cert *tls.Certificate
+}
+
+func NewCertReloader(certPath, keyPath string) (*CertReloader, error) {
+	watchables := make(map[string]string)
+
+	w, in, err := newWatchable(certPath)
+	if err != nil {
+		return nil, fmt.Errorf("error adding cert watchable: %v", err)
+	}
+	watchables[w] = in
+
+	w, in, err = newWatchable(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("error adding key watchable: %v", err)
+	}
+	watchables[w] = in
+
+	r := &CertReloader{
+		certPath:   certPath,
+		keyPath:    keyPath,
+		watchables: watchables,
+	}
+
+	if err := r.reload(); err != nil {
+		return nil, fmt.Errorf("error loading certificates: %v", err)
+	}
+
+	return r, nil
+}
+
+func newWatchable(target string) (watchable, in string, _ error) {
+	// simple case: the target is a file.
+	// in that case we watch for changes in its own directory.
+	watchable = filepath.Clean(target)
+	in, _ = filepath.Split(target)
+	in = filepath.Clean(in)
+
+	stat, err := os.Lstat(target)
+	if err != nil {
+		return "", "", fmt.Errorf("lstat on %q failed: %v", target, err)
+	}
+
+	// k8s case: the target is a symlink.
+	// Here, we watch the intermediate symlink (named `..data`)
+	// in the same directory as the target.
+	if stat.Mode()&os.ModeSymlink != 0 {
+		dest, err := os.Readlink(target)
+		if err != nil {
+			return "", "", fmt.Errorf("lstat on %q failed: %v", target, err)
+
+		}
+
+		watchable, _ = filepath.Split(dest)
+		watchable = filepath.Clean(watchable)
+	}
+
+	return
+}
+
+// Watch watches the configured certificate and key path and blocks the current goroutine
+// until the scenario context is done or an error occurred during reloading.
+func (r *CertReloader) Watch(ctx context.Context) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+
+	for _, v := range r.watchables {
+		glog.V(4).Infof("watching: %q", v)
+
+		if err := watcher.Add(v); err != nil {
+			return fmt.Errorf("error adding watchable: %v", err)
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case event := <-watcher.Events:
```

Comment:
I wonder if we, in addition should add a ticker anyways. I am not sure we have 100% guarantees to catch all file system events, cc @lucab.

Reply:
Curious: for this very case, why not just read the files once per minute and apply if they changed? For other cases I agree that reloading as quickly as possible when available makes perfect sense, but here we have a lot of time to apply cert rotation anyways no?

Reply:
sounds good yes, this will simplify a lot of things and we don't have to think about edge cases :+1: let me implement a ticker-only based approach here and reuse the implemented logic here for prometheus-config-reloader.